In [7]:
import pandas as pd
import numpy as np
df = pd.DataFrame(np.random.random((5,4)),columns=['c0','c1','c2','c3'],index=['r0','r1','r2','r3','r4'])
df.to_markdown()

'|    |        c0 |       c1 |       c2 |       c3 |\n|:---|----------:|---------:|---------:|---------:|\n| r0 | 0.169489  | 0.226653 | 0.35706  | 0.199845 |\n| r1 | 0.181289  | 0.289734 | 0.392262 | 0.228899 |\n| r2 | 0.0492157 | 0.429355 | 0.299516 | 0.566478 |\n| r3 | 0.304988  | 0.149374 | 0.801296 | 0.699043 |\n| r4 | 0.236938  | 0.652044 | 0.132635 | 0.519792 |'

# Pandas
### Overview
To be able to use pandas you need to import it into the current python scope. It is common to import pandas using the alias pd becaue it makes calling it more succinct using the `import <module> as <alias>`. 

```Python
import pandas as pd
```
Pandas has a custom data structure called a `DataFrame` which is a very powerful multidmensional, hetrogeneous data type storer. The data are stored using an index and a column name. For example, in the example below the indexes are r0-r4 and columns are c0-c3. 

|    |        c0 |       c1 |       c2 |       c3 |
|:---|----------:|---------:|---------:|---------:|
| r0 | 0.169489  | 0.226653 | 0.35706  | 0.199845 |
| r1 | 0.181289  | 0.289734 | 0.392262 | 0.228899 |
| r2 | 0.0492157 | 0.429355 | 0.299516 | 0.566478 |
| r3 | 0.304988  | 0.149374 | 0.801296 | 0.699043 |
| r4 | 0.236938  | 0.652044 | 0.132635 | 0.519792 |


#### Creating a DataFrame
```Python
import pandas as pd
import numpy as np
df = pd.DataFrame(np.random.random((5,4)),columns=['c0','c1','c2','c3'],index=['r0','r1','r2','r3','r4'])
```

#### Importing from files
Pandas can import data from many different regularly formatted datasets
* When importing from a csv file you can specify:
    * the delimiter - e.g. comma, semicolon, tab, space
    * the header names for the column names
    * Column index for the row index names, default is None and these are just assigned to be indexed 0-Nrows
```Python
import pandas as pd
df = pd.read_csv(filename,sep=',',header,index_col=None)
```
* Importing from excel requires
    * sheet_name
    * header row index
    * index_col
```Python
import pandas as pd
df = pd.read_excel(filename,sheet_name,header,index_col=None)
```
* Python datasets and objects can also be saved in a binary format using Pickel. It is possible to save the files using this format as well. 
```Python
import pandas as pd
df = pd.read_pickle(filepath)
```

**Exercise**
1. Create a pandas DataFrame 
2. Load a csv file into pandas 
3. Load an excel spreadsheet

### Selecting data by column and row
Using the same DataFrame as before:

|    |        c0 |       c1 |       c2 |       c3 |
|:---|----------:|---------:|---------:|---------:|
| r0 | 0.169489  | 0.226653 | 0.35706  | 0.199845 |
| r1 | 0.181289  | 0.289734 | 0.392262 | 0.228899 |
| r2 | 0.0492157 | 0.429355 | 0.299516 | 0.566478 |
| r3 | 0.304988  | 0.149374 | 0.801296 | 0.699043 |
| r4 | 0.236938  | 0.652044 | 0.132635 | 0.519792 |

We can access the elements inside the DataFrame by either:
* using the column name and the index: `df.loc[index,column_name]`
```Python
df.loc['r1','c3']
0.2288986004071356
```
* Or we can access the same value using the column index and row index `df.iloc[rowindex,colindex]`
```Python
df.iloc[1,3]
0.2288986004071356
```
We can also use this syntax to update the values of the DataFrame and combine it with slicing
```Python
df.iloc[1,:] = 0
df.loc['r3',:] = 10
```

This would update the previous dataframe to be

|    |        c0 |       c1 |       c2 |       c3 |
|:---|----------:|---------:|---------:|---------:|
| r0 | 0.169489  | 0.226653 | 0.35706  | 0.199845 |
| r1 | 0.000000  | 0.000000 | 0.000000 | 0.000000 |
| r2 | 0.0492157 | 0.429355 | 0.299516 | 0.566478 |
| r3 | 10.000000  | 10.000000 | 10.000000 | 10.000000 |
| r4 | 0.236938  | 0.652044 | 0.132635 | 0.519792 |

**Exercise**

Using the following DataFrame
```Python
import pandas as pd
df = pd.read_csv("example_data_frame.csv")
```
Create a dataframe with a row index that has a keyword and a number. Use a boolean mask to update only the values of different experiments to be `**2` and others to be normalised and some can be standardised
```Python
def normalise(array):
    '''
    Make a dataset range from 0 to 1
    '''
    minval = np.min(array)
    maxval = np.max(array)
    array-=minval
    array/=(maxval-minval)
```

1. Update every second row of the DataFrame to be squared
2. 

In [16]:
# df.loc['r1','c3']
df.iloc[1,:] = 0
df.loc['r3',:] = 10
df

c0         c1         c2         c3
r0   0.169489   0.226653   0.357060   0.199845
r1   0.000000   0.000000   0.000000   0.000000
r2   0.049216   0.429355   0.299516   0.566478
r3  10.000000  10.000000  10.000000  10.000000
r4   0.236938   0.652044   0.132635   0.519792

### Slicing data frame
We can extract only the data for one column by slicing the dataframe with the column name e.g. `dataframe['col1']` would return a pandas series for just a single column. 

If you want to extract multiple columns we can do this by passing a list of column headers e.g. `dataframe[['col1','col2']]` which returns a copy of the original dataframe with only `col1` and `col2`.

**Note: Slicing DataFrames creates a copy of the DataFrame, and if the DataFrame is updated the using the copy the original DataFrame won't be changed**

### Adding columns and rows to a DataFrame
Adding a new column to a DataFrame is very easy, it is similar to adding a new item to a dictionary where the column index is the key and the column data is the item. 

```Python
dataframe['newcolumn'] = values
```

Adding a new row to the DataFrame can be done by inserting a new index into the DataFrame.
```Python
dataframe.loc['newrowindex'] = rowvalues
```

### Deleting columns and rows
We can delete a column from the DataFrame using the `drop()` method:
```Python
df.drop('c2',axis=1,inplace=False) # return a new df object without column 'c2'
df.drop('c2',axis=1,inplace=True) # update the current df object to remove 'c2'
```

In the same way rows can be deleted setting the axis argument to 0
```Python
df.drop('r2',axis=0,inplace=False) # return a new df object without row 'r2'
df.drop('r2',axis=0,inplace=True) # update the current df object to remove row 'r2'
```



### Rename columns or indexes in the DataFrame
Column names in the DataFrame can be updated using the rename method
```Python
column_names = {'c0' : 'new_c0',
                'c1' : 'new_c1',
                'c2' : 'new_c2',
                'c3' : 'new_c3'
               }
df.rename(columns=column_names)
```

### Exporting data frames
DataFrames can be exported to many different formats using Pandas in built functions. 
* The contents of a numerical DataFrame can be expored into a numpy array using `to_numpy()`
* The DataFrame can be exported into pythons binary fomart using `to_pickle(filename)`
* The DataFrame can be exported into a csv using `to_csv(filename)` many of the arguments for importing csv's such as the delimiter and header can be specified as additional arguments.
* The DataFrame can be exported into an excel worksheet using `to_excel(filename)`. If you want to export multiple DataFrames to the same spreadsheet a unique `worksheet` needs to be defined in as an argument in the function. 

